### PR TITLE
systemd/services: Rewrite Service into a functional component

### PR DESCRIPTION
**Pull request by martinpitt**
After cockpit-project/cockpit#18422, I want to address the remaining state handling instabilities like cockpit-project/cockpit#18465. I started with the details page, that's a bit easier to do.

 - Goes on top of #18494

### CODE REVIEW

**Comment by martinpitt** [comment]
[these](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-054204-40c59661-centos-8-stream/log.html#299-2) [three](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-054204-40c59661-centos-8-stream-pybridge/log.html#299-2) [failures](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-054204-40c59661-fedora-37-pybridge/log.html#299), it missed a state update.

This [already happens on main](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18480-20230310-064510-601cd201-rhel-8-8/log.html#271)

[this failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-054204-40c59661-rhel-8-8/log.html#299-2) is different: "object path is invalid in dbus "call": (null)".

I cannot reproduce these locally. This needs debug enabling and log staring.

File: pkg/systemd/services/service.jsx [hunk 1]
```diff
@@ -40,147 +41,140 @@ function debug() {
...
+    const [path, setPath] = useState(null);
+    const [unitProps, setUnitProps] = useState(null);
+
+    const updateProperties = () => {
+        const promises = [dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_UNIT])];
+
+        if (unitId.endsWith(".timer"))
+            promises.push(dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_TIMER]));
+        else if (unitId.endsWith(".socket"))
+            promises.push(dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_SOCKET]));
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:56]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-072502-c32f9a69-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 2]
```diff
@@ -40,147 +41,140 @@
...
+                    const other_props = replies[1]?.[0];
+
+                    // unwrap variants
+                    for (const key in unit_props)
+                        unit_props[key] = unit_props[key].v;
+
+                    if (unitId.endsWith(".timer"))
+                        addTimerProperties(other_props, unit_props);
+                    else if (unitId.endsWith(".socket"))
+                        unit_props.Listen = other_props.Listen.v;
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:70]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-072502-c32f9a69-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 3]
```diff
@@ -40,147 +41,140 @@
...
+    };
+
+    useObject(
+        () => {
+            dbusClient.call(s_bus.O_MANAGER, s_bus.I_MANAGER, "LoadUnit", [unitId])
+                    .then(([path]) => {
+                        debug("Service detail", unitId, "loaded path", path);
+                        setPath(path);
+                    })
+                    .catch(ex => setError(ex.toString()));
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:88]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-072502-c32f9a69-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 4]
```diff
@@ -40,147 +41,140 @@
...
-            </WithDialogs>
-        );
-    }
-}
+            }),
+        sub => sub.remove(), [path]);
+
+    // render
+    if (error)
+        return <EmptyStatePanel title={_("Loading unit failed")} icon={ExclamationCircleIcon} paragraph={error} />;
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:131]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230310-072502-c32f9a69-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx (hunk 1 above)
Comment (martinpitt) [reply pkg/systemd/services/service.jsx:56]: Hmm, that is odd. testBasic does check the details page of `dbus.socket`, and ensures it shows a correct "Listening on" path.

File: pkg/systemd/services/service.jsx (hunk 3 above)
Comment (martinpitt) [reply pkg/systemd/services/service.jsx:88]: Correct, annotated.

File: pkg/systemd/services/service.jsx (hunk 4 above)
Comment (martinpitt) [reply pkg/systemd/services/service.jsx:131]: Correct, annotated.

File: pkg/systemd/services/service.jsx (hunk 1 above)
Comment (martinpitt) [reply pkg/systemd/services/service.jsx:56]: Oh, but not for an update!

File: pkg/systemd/services/service.jsx [hunk 5]
```diff
@@ -40,147 +41,143 @@ function debug() {
...
+export const Service = ({ dbusClient, owner, unitId, unitIsValid, addTimerProperties, pinnedUnits }) => {
+    const _path = useRef(null);
+    const [error, setError] = useState(null);
+    const [reloading, setReloading] = useState(false);
+    const [unitProps, setUnitProps] = useState(null);
+
+    const updateProperties = () => {
+        const path = _path.current?.path;
+        if (!path)
+            console.error("Service detail", unitId, "updateProperties called without path");
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:53]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-094909-6117a6e5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 6]
```diff
@@ -40,147 +41,143 @@
...
+        const promises = [dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_UNIT])];
+
+        if (unitId.endsWith(".timer"))
+            promises.push(dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_TIMER]));
+        else if (unitId.endsWith(".socket"))
+            promises.push(dbusClient.call(path, s_bus.I_PROPS, "GetAll", [s_bus.I_SOCKET]));
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:59]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-094909-6117a6e5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 7]
```diff
@@ -40,147 +41,143 @@
...
+                    const other_props = replies[1]?.[0];
+
+                    // unwrap variants
+                    for (const key in unit_props)
+                        unit_props[key] = unit_props[key].v;
+
+                    if (unitId.endsWith(".timer"))
+                        addTimerProperties(other_props, unit_props);
+                    else if (unitId.endsWith(".socket"))
+                        unit_props.Listen = other_props.Listen.v;
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:73]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-094909-6117a6e5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 8]
```diff
@@ -40,147 +41,143 @@
...
+
+    // load object path and set up PropertiesChanged subscription whenever unitId changes
+    useObject(
+        () => {
+            dbusClient.call(s_bus.O_MANAGER, s_bus.I_MANAGER, "LoadUnit", [unitId])
+                    .then(([path]) => {
+                        debug("Service detail", unitId, "loaded path", path);
+
+                        if (_path.current?.propSub)
+                            _path.current.propSub.remove();
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:92]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-094909-6117a6e5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:53]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-105853-18805bab-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:59]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-105853-18805bab-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx (hunk 7 above)
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:73]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-105853-18805bab-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx (hunk 8 above)
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:92]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-105853-18805bab-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 9]
```diff
@@ -40,147 +41,141 @@ function debug() {
...
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:90]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-144841-38fcc860-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (cockpituous) [inline pkg/systemd/services/service.jsx:90]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-161204-a7c3ee57-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/services/service.jsx [hunk 10]
```diff
@@ -40,147 +41,137 @@ function debug() {
...
+    };
+
+    // load object path and set up PropertiesChanged subscription whenever unitId changes
+    useObject(
+        () => {
+            dbusClient.call(s_bus.O_MANAGER, s_bus.I_MANAGER, "LoadUnit", [unitId])
+                    .then(([path]) => {
+                        debug("Service detail", unitId, "loaded path", path);
+
+                        if (_path.current?.propSub)
```
Comment (KKoukiou) [inline pkg/systemd/services/service.jsx:89]: This is very nicely replace by opt chaining, _path.current?.propSub?.remove() and we can drop the condition.

**Review by KKoukiou — APPROVED** [review]

File: pkg/systemd/services/service.jsx (hunk 10 above)
Comment (martinpitt) [reply pkg/systemd/services/service.jsx]: You are right! We don't even need the second ?, as propSub always gets set. Updated, thanks for pointing out!

**Review by KKoukiou — APPROVED** [review]

File: pkg/systemd/services/service.jsx [hunk 11]
```diff
@@ -40,147 +41,139 @@ function debug() {
...
+    };
+
+    // load object path and set up PropertiesChanged subscription whenever unitId changes
+    useObject(
+        () => {
+            dbusClient.call(s_bus.O_MANAGER, s_bus.I_MANAGER, "LoadUnit", [unitId])
+                    .then(([path]) => {
+                        debug("Service detail", unitId, "loaded path", path);
+
+                        _path.current?.propSub.remove();
```
Comment (cockpituous) [inline pkg/systemd/services/service.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18481-20230313-183150-83c26568-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Pull request merged**